### ISSUE TITLE
Modify _add_docstr to also set the correct module for the APIs

### ISSUE THREAD

**Issue by anjali411** [pytorch/pytorch#80377]
### 📚 The doc issue

Currently, many public APIs don't have their module set correctly to meet our requirements for public API (https://github.com/pytorch/pytorch/wiki/Public-API-definition-and-documentation). All the ops in `torch.linalg`, `torch.special`, `torch.sparse` etc currently use `_add_docstr` to correctly set the doc for all the functions in these modules. The proposal here is to modify `_add_docstr` to set the correct module. 

Once this is done, we can get rid of all the respective entries in this allowlist: https://github.com/pytorch/pytorch/blob/d67ce755ad1439b03d4abcc2d40496b20cfd570e/test/allowlist_for_publicAPI.json#L1786-L1827

**Comment by billdarrow** [comment]
@anjali411 is it sufficient to correct the module name as follows, `torch._C._linalg` -> strip('_C._') -> `torch.linalg`? and the idea is it becomes a new side-effect of _add_docstr?

**Comment by anjali411** [comment]
@billdarrow I don't think so! I updated the module name for the APIs in linalg and special in this PR: https://github.com/pytorch/pytorch/pull/81126 but currently, their `__name__` attribute has an incorrect value as well.

**Comment by billdarrow** [comment]
ack I see
